galois embedded crypto module and secure transport

podhrmic opened the pull request:
Based on discussion about secure link here https://github.com/paparazzi/paparazzi/pull/2201
I refactored and moved the code in paparazzi per @gautierhattenberger suggestions.

Current status - gec transport behaves like a transparent layer and does nothing special (so we have a regular transparent pprzlink). Crypto stuff to come soon.

This PR replaces #2201

podhrmic commented:
Of course, the build now fails, because Travis or Semaphore doesn't know about Rust.

podhrmic commented:
Status update: the crypto transport is not yet functional, but you can see where I am heading to.

gautierhattenberger commented on sw/airborne/modules/datalink/gec_dl.c:
```diff
@@ -0,0 +1,382 @@
+  msg.trans = trans;
+  msg.dev = dev;
+  msg.sender_id = AC_ID;
+  msg.receiver_id = PPRZLINK_MSG_BROADCAST;
```
This is only needed if you want to broadcast the message to all aircraft around you and not only the GCS. If only the GCS should receive the message, using `pprz_msg_send_SECURE_LINK_STATUS` is fine and will work for both protocol.

podhrmic replied:
Fixed.

podhrmic commented on Makefile.ac:
```diff
@@ -187,6 +188,7 @@ flight_plan_ac_h : $(FLIGHT_PLAN_H) $(FLIGHT_PLAN_XML)
 
 makefile_ac: $(MAKEFILE_AC)
 
+ifeq ($(GENERATE_KEYS),1)
```
Can you please elaborate? Even better if you could outline the solution, I am not sure if I understand enough makefile magic.

gautierhattenberger replied:
see https://github.com/paparazzi/paparazzi/tree/gec_module_makefile
if it fails afterward because no keys are available, maybe we can provide dummy keys for compilation purpose

podhrmic replied:
Great, thanks! The easy way is to remove the airframe with secure link from tests, but that is not a sustainable solution. Can we add Rust on Travis?

gautierhattenberger replied:
around here I guess: https://github.com/paparazzi/paparazzi/blob/master/.travis.yml#L12

podhrmic replied:
Fixed

podhrmic commented on sw/airborne/modules/datalink/gec/gec.c:
```diff
@@ -0,0 +1,127 @@
+ */
+void generate_ephemeral_keys(struct gec_privkey *sk)
+{
+  for (uint8_t i = 0; i < (PPRZ_KEY_LEN / sizeof(uint32_t)); i++) {
```
Changed to uint16 for peace of mind

podhrmic commented on sw/airborne/modules/datalink/gec/gec.h:
```diff
@@ -0,0 +1,130 @@
+#include "std.h"
+#include "mcu_periph/rng.h"
+#include "generated/keys_uav.h"
+#include "../ext/hacl-c/Hacl_Ed25519.h"
```
I am getting "No such file or directory" error with the include, it doesn't seem to work (although the include directory is properly added in the makefile)

podhrmic replied:
`module.dtd` doesn't contain `include` tag - perhaps this is some new change? Works as is so I would leave it for now.

gautierhattenberger replied:
strange, it should be available in master: 
https://github.com/paparazzi/paparazzi/blob/master/conf/modules/module.dtd#L89
and used like this:
https://github.com/paparazzi/paparazzi/blob/master/conf/modules/cv_qrcode.xml

podhrmic commented:
OK, I don't know how to make Travis happy. Here are the installation instructions: https://wiki.paparazziuav.org/wiki/Pprzlink#Installation

The problem is that hacl-c has to be compiled into a library, and then copied into `/usr/local/lib` (requires root) so cargo can find it for the key generator. Similar problem exists for Rust link - both programs need to have the hacl library in systems directory.

gautierhattenberger commented:
there is really no way to tell rest to look a other places than the system ?

gautierhattenberger commented:
maybe the `rustc-flags` with the proper `-L` flag ? (http://doc.crates.io/build-script.html)

podhrmic commented:
yep, the build script is a way to go. I ll update the rust modules.

podhrmic commented:
The reason Travis doesn't compile is that it tries to compile hacl library on the host machine with `arm-none-eabi-gcc` instead of regular `gcc` (see https://travis-ci.org/paparazzi/paparazzi/jobs/324390713#L3864 for full output). The actual command is invoked from `rusthacl` 's build script (here https://github.com/podhrmic/rusthacl/blob/master/build.rs#L14). The hacl makefile uses `$(CC)` to invoke the compiler. All works nicely on my local machine.

@flixr @gautierhattenberger any ideas what can be wrong with Travis and how to fix it?

gautierhattenberger commented:
I'm really not good with Travis config, maybe @flixr will have an idea

flixr commented:
While I don't have any idea why this fails on Travis, you can also restart the build in the Travis web interface... did you change anything that I missed which might make it succeed?

podhrmic commented:
I think I am getting closer with manually specifying the value of '$(CC)' for the rusthacl build.

podhrmic commented on conf/modules/haclc.xml:
```diff
@@ -19,6 +20,7 @@
     <raw>
       # to not use 128 bit arithmetic
       ap.CFLAGS += -DKRML_NOUINT128
```
The `GENERATE_KEYS` flag has to be in `<raw>`, otherwise it doesn't trigger the generation.

podhrmic commented:
@gautierhattenberger I addressed your comments and fixed the build process. Now the majority of work consists of improving the rustlink process to neatly handle encryption, and in improving the gec module itself.

Would be helpful to merge a minimal working example and improve it incrementally, or do you prefer to polish and properly test this PR first and then merge? Since there is not that much happening in master right now, I am not too worried about rebasing.

gautierhattenberger commented:
can you rebase or merge to solve conflicts with master ?

podhrmic commented:
Which conflicts? Github says there are no conflicts with master.

gautierhattenberger commented:
I can't find the pprzlink commit. Are you sure you have checked out the correct one from latest pprzlink/master ?

podhrmic commented:
It was merged into master, see https://github.com/paparazzi/pprzlink/pull/73
I updated it to point at the latest pprzlink master just now.
